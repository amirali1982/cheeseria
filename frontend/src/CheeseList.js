import React, {Component} from 'react';
import {Button, ButtonGroup, Container, Table} from 'reactstrap';
import AppNavbar from './AppNavbar';
import EstimatedCosts from './components/EstimatedCosts'
import {Link} from 'react-router-dom';

class CheeseList extends Component {

    constructor(props) {
        super(props);
        this.state = {cheeseria: [], totalEstimatedCost: 0};
        this.remove = this.remove.bind(this);
    }

    componentDidMount() {
        fetch('/cheeseria')
            .then(response => response.json())
            .then(data => this.setState({cheeseria: data}))
    }

    async remove(id) {
        await fetch(`/cheeseria/${id}`, {
            method: 'DELETE',
            headers: {
                'Accept': 'application/json',
                'Content-Type': 'application/json'
            }
        }).then(() => {
            let updatedCheeseria = [...this.state.cheeseria].filter(i => i.id !== id);
            this.setState({cheeseria: updatedCheeseria});
        });
    }

    handleReqWeightChange = (event, id) => {
        let totalEstimatedCost = 0;
        for (const cheese of this.state.cheeseria) {
            if (cheese.id === id) {
                cheese.estCost = event.target.value * cheese.price
            }
            if (cheese.estCost != null) {
                totalEstimatedCost = totalEstimatedCost + cheese.estCost
            }
        }
        this.setState({totalEstimatedCost: totalEstimatedCost}, () => {
            console.log(this.state.totalEstimatedCost);
        });
    }

    render() {
        const {isLoading} = this.state;

        if (isLoading) {
            return <p>Loading...</p>;
        }

        const cheeseList = this.state.cheeseria.map(cheese => {
            return <tr key={cheese.id}>
                <td><img src={`data:image/jpeg;base64,${cheese.image}`} alt={cheese.name}/></td>
                <td style={{whiteSpace: 'nowrap'}}>{cheese.name}</td>
                <td>{cheese.color}</td>
                <td>{cheese.price}</td>
                <td><input type={"text"} id={"reqWeight-" + cheese.id}
                           onChange={event => this.handleReqWeightChange(event, cheese.id)}/>
                </td>
                <td>
                    <ButtonGroup>
                        <Button size="sm" color="primary" tag={Link} to={"/cheeseria/" + cheese.id}>Edit</Button>
                        <Button size="sm" color="danger" onClick={() => this.remove(cheese.id)}>Delete</Button>
                    </ButtonGroup>
                </td>
            </tr>
        });

        return (
            <div>
                <AppNavbar/>
                <Container fluid>
                    <br/>
                    <h2>Cheeseria</h2>
                    <Table className="mt-4">
                        <thead>
                        <tr>
                            <th width="20%">Image</th>
                            <th width="20%">Name</th>
                            <th width="10%">Color</th>
                            <th width="10%">Price/kg</th>
                            <th width="10%">Req. Weight</th>
                            <th width="10%">Actions</th>
                        </tr>
                        </thead>
                        <tbody>
                        {cheeseList}
                        </tbody>
                    </Table>
                    <EstimatedCosts value={this.state.totalEstimatedCost}/>
                </Container>
            </div>
        );
    }
}

export default CheeseList;