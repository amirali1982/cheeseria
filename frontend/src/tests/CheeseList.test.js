import renderer from 'react-test-renderer';
import {MemoryRouter} from "react-router-dom";
import CheeseList from "../CheeseList";

it('Test Cheese List', () => {
    const component = renderer.create(
        <MemoryRouter>
            <CheeseList />
        </MemoryRouter>
    )

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

})