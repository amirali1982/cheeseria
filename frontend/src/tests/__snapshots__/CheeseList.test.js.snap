// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Cheese List 1`] = `
<div>
  <nav
    className="navbar navbar-expand-md navbar-dark bg-dark"
  >
    <a
      className="navbar-brand"
      href="/cheeseria-ui"
      onClick={[Function]}
    >
      Home
    </a>
    <div
      className="float-right"
    >
      <a
        aria-label={null}
        className="btn btn-success"
        href="/cheeseria-ui/new"
        onClick={[Function]}
      >
        Add Cheese
      </a>
    </div>
  </nav>
  <div
    className="container-fluid"
  >
    <br />
    <h2>
      Cheeseria
    </h2>
    <div>
      <h4>
        Estimated Cost: 
        0
      </h4>
    </div>
    <table
      className="mt-4 table"
    >
      <thead>
        <tr>
          <th
            width="20%"
          >
            Image
          </th>
          <th
            width="20%"
          >
            Name
          </th>
          <th
            width="10%"
          >
            Color
          </th>
          <th
            width="10%"
          >
            Price/kg
          </th>
          <th
            width="10%"
          >
            Req. Weight
          </th>
          <th
            width="10%"
          >
            Actions
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
    <div>
      <h4>
        Estimated Cost: 
        0
      </h4>
    </div>
  </div>
</div>
`;
